geode-core/src/distributedTest/java/org/apache/geode/internal/cache/partitioned/fixed/FixedPartitioningWithTransactionDistributedTest.java

```diff
@@ -188,6 +193,100 @@ public void clientExecuteFunctionOnMovedPrimaryBucketFailWithTransactionDataReba
         .isInstanceOf(TransactionDataRebalancedException.class);
   }
 
+  @Test
+  public void clientCanRollbackFunctionOnRegionWithFilterAndWithSingleHopEnabled() {
+    setupServers();
+    setupClient();
+
+    Region region = clientCacheRule.getClientCache().getRegion(regionName);
+    CacheTransactionManager txManager =
+        clientCacheRule.getClientCache().getCacheTransactionManager();
+
+    TransactionId transactionId =
+        doFunctionTransactionAndSuspend(region, txManager, new MyTransactionFunction());
+    txManager.resume(transactionId);
+    txManager.rollback();
+
+    server1.invoke(() -> {
+      assertThat(cacheRule.getCache().getRegion(regionName).get(2)).isEqualTo(2);
+    });
+  }
+
+  @Test
+  public void clientCanRollbackFunctionOnRegionWithoutFilterAndWithSingleHopEnabled() {
+    setupServers();
+    setupClient();
+
+    Region region = clientCacheRule.getClientCache().getRegion(regionName);
+    CacheTransactionManager txManager =
+        clientCacheRule.getClientCache().getCacheTransactionManager();
+
+    try {
+      TransactionId transactionId = doFunctionTransactionAndSuspend(region, txManager,
+          new MyTransactionFunction(), Type.ON_REGION, false);
+      txManager.resume(transactionId);
+      txManager.rollback();
+    } catch (FunctionException functionException) {
```

Comment:
You should either add a "fail" statement in before the catch-block or change this to use AssertJ's catchThrowable:
```
import static org.assertj.core.api.Assertions.catchThrowable;

Throwable thrown = catchThrowable(() -> txManager.rollback());
assertThat(thrown).isInstanceOf(FunctionException.class);
assertThat(thrown.getCause()).isInstanceOf(ServerOperationException.class);
assertThat(thrown.getCause().getCause()).isInstanceOf(FunctionException.class);
assertThat(thrown.getCause().getCause().getCause()).isInstanceOf(TransactionDataRebalancedException.class);
etc
```

Reply:
The issue here is that doFunctionTransactionAndSuspend() may or may not fail with FunctionException. The test has a function without filter, so it can land to any server -- if the server has the primary bucket, the transaction with function will be successful; if the server receive the request does not have the primary bucket, it will throw the transaction exception wrapped in function exception.
These two scenario are both valid. The only verification is that after function the entry will not be removed/destroyed.